### DISCUSSION TITLE
Access to Virtual Authenticators

### DISCUSSION THREAD

**Discussion by Jwashton** [keathley/wallaby#700] [Q&A] [answered]
I'd like to test a site that uses WebAuthn for authentication. The WebAuthn spec [extends the WebDriver API](https://w3c.github.io/webauthn/#sctn-automation) to include calls for creating, removing, and querying virtual authenticators and credentials.

How can I make use of these features in Wallaby? I don't see functions in Wallaby's documentation for virtual authenticators specifically. Is there an escape hatch that would let me define custom requests to the WebDriver API? Maybe a lower-level call I can make?

**Comment by mhanberg** [comment, marked as answer]
I would need to personally research the WebDriver extension specification, but if you can create extensions by the HTTP interface, then you should be able to accomplish this in your code in the meantime.

You can use the function  `Wallaby.HTTPClient.request(method, url, params \\ %{}, opts \\ [])` to send requests to the webdriver. 

`method` is the http method, and `url` is the webdrivers server url. This can be fetched from the current session `session.url`.

**Reply by Jwashton** [reply]
That looks like exactly what I need. Thank you!

**Reply by mhanberg** [reply]
Please share your success/failures on this, I'm sure other Wallaby users would find it useful and maybe something that can be upstreamed (if there is the right fit).
